opal/util/opal_gethostname.c

```diff
@@ -0,0 +1,173 @@
+
+
+#define NUM_TRIES_FOR_NULL_HOSTNAME 8
+
+#define MEMORY_ERROR_STRING "Invalid address (EFAULT) in call to gethostname. Check memory usage."
+#define HOSTNAMELEN_ERROR_STRING "Error in call to gethostname() - check hostname length. Excessively long hostnames are unsupported."
+#define UNKNOWN_ERROR_STRING "Unknown error in opal_gethostname() function. Probable code issue."
+/*
+ * This gethostname wrapper does not return the full-length hostname in
+ * those cases where it is too long for the buffer. It does, however,
+ * guarantee a null-terminated hostname is returned, even if it's
+ * truncated. It also tries again in the case where gethostname returns an 
+ * error because the buffer is too short, but once it has retrieved the full
+ * name without error, it truncates the result and resets errno to 
+ * ENAMETOOLONG. 
+ */
+static char opal_local_hostname[OPAL_LOCAL_MAXHOSTNAMELEN] =
+    HOSTNAMELEN_ERROR_STRING;
+
+
+const char *opal_gethostname( void )
+{
+    size_t count;
+    size_t length = OPAL_LOCAL_MAXHOSTNAMELEN;
+    int ret_val, num_tries = 0;
+    int temp_errno = 0;
+
+    if( first_pass ) {
+        char *buf = malloc( length );
+        first_pass = false;
+        /*
+         * Terminate the buffer first in the 0th position in case we get a 
+         * null result.
+         */
+        buf[0] = '\0';
+
+        while( 1 ) {
+            /*
+             * Offer all but the last byte of the buffer to gethostname. 
+             */
+            ret_val = gethostname( buf, length - 1 );
+            /*
+             * Terminate the buffer in the last position.
+             */
+            buf[length - 1] = '\0';
+            if( 0 == ret_val ) {
+                count = strlen( buf );
+                /*The result was not truncated */
+                if( count > 0 && count < length - 1 ) {
+                    break;
+                }
+                /*
+                 * The buffer is empty. In some gethostname
+                 * implementations this can be because the buffer was too 
+                 * small. But, if it is zero for some other reason, we
+                 * could get caught in the loop. So, break after
+                 * NUM_TRIES_FOR_NULL_HOSTNAME. 
+                 */
+                else if( 0 == count ) {
+                    if( num_tries > NUM_TRIES_FOR_NULL_HOSTNAME ) {
+                        ret_val = -1;
+                        temp_errno = EINVAL;
+                        break;
+                    }
+                    num_tries++;
+                }
+                /* The result *may* be truncated (can't know for sure, so we increase 
+                 * the length of the buffer (happens below) and try again). */
+                else if( count == length - 1 ) {
+			; /* no op */
+                }
+                /* Unidentified case, implies some bug in this code or some unknown
+                 * return string from gethostname() */
+                else {
```

Comment:
We're in the success block (IE - a return value of 0) - so I'm not sure what case this might cover.

My understanding is if it returns 0, it succeeded - with a possibility that the hostname may have been truncated.

Reply:
See above. I am trying to account for all possible behaviors of `gethostname()`, even ones that we would not expect.